## About

The dataset containing 3 months’ worth of transactions for 100 hypothetical customers. It contains purchases, recurring transactions, and salary transactions.

## Objective

- start by doing some basic checks – are there any data issues? Does the data need to be cleaned?

- Gather some interesting overall insights about the data. For example -- what is the average transaction amount? How many transactions do customers make each month, on average?

- Segment the dataset by transaction date and time. Visualise transaction volume and spending over the course of an average day or week. Consider the effect of any outliers that may distort your analysis.

- For a challenge – what insights can you draw from the location information provided in the dataset?

- Put together 2-3 slides summarising your most interesting findings to ANZ management.

## EDA

In [1]:
#import required packages
import pandas as pd
print("....pandas imported")
print("all packages imported")

....pandas imported
all packages imported


In [2]:
master_df = pd.read_excel("ANZ synthesised transaction dataset.xlsx")

In [3]:
df = master_df.copy()

In [4]:
pd.set_option('display.max_columns', None)
df.head()

status  card_present_flag bpay_biller_code         account currency  \
0  authorized                1.0              NaN  ACC-1598451071      AUD   
1  authorized                0.0              NaN  ACC-1598451071      AUD   
2  authorized                1.0              NaN  ACC-1222300524      AUD   
3  authorized                1.0              NaN  ACC-1037050564      AUD   
4  authorized                1.0              NaN  ACC-1598451071      AUD   

        long_lat txn_description                           merchant_id  \
0  153.41 -27.95             POS  81c48296-73be-44a7-befa-d053f48ce7cd   
1  153.41 -27.95       SALES-POS  830a451c-316e-4a6a-bf25-e37caedca49e   
2  151.23 -33.94             POS  835c231d-8cdf-4e96-859d-e9d571760cf0   
3  153.10 -27.66       SALES-POS  48514682-c78a-4a88-b0da-2d6302e64673   
4  153.41 -27.95       SALES-POS  b4e02c10-0852-4273-b8fd-7b3395e32eb0   

   merchant_code first_name  balance       date gender  age merchant_suburb  \
0            NaN      Diana    35.39 2018-08-01      F   26         Ashmore   
1            NaN      Diana    21.20 2018-08-01      F   26          Sydney   
2            NaN    Michael     5.71 2018-08-01      M   38          Sydney   
3            NaN     Rhonda  2117.22 2018-08-01      F   40         Buderim   
4            NaN      Diana    17.95 2018-08-01      F   26   Mermaid Beach   

  merchant_state                    extraction  amount  \
0            QLD  2018-08-01T01:01:15.000+0000   16.25   
1            NSW  2018-08-01T01:13:45.000+0000   14.19   
2            NSW  2018-08-01T01:26:15.000+0000    6.42   
3            QLD  2018-08-01T01:38:45.000+0000   40.90   
4            QLD  2018-08-01T01:51:15.000+0000    3.25   

                     transaction_id    country     customer_id  \
0  a623070bfead4541a6b0fff8a09e706c  Australia  CUS-2487424745   
1  13270a2a902145da9db4c951e04b51b9  Australia  CUS-2487424745   
2  feb79e7ecd7048a5a36ec889d1a94270  Australia  CUS-2142601169   
3  2698170da3704fd981b15e64a006079e  Australia  CUS-1614226872   
4  329adf79878c4cf0aeb4188b4691c266  Australia  CUS-2487424745   

  merchant_long_lat movement  
0     153.38 -27.99    debit  
1     151.21 -33.87    debit  
2     151.21 -33.87    debit  
3     153.05 -26.68    debit  
4     153.44 -28.06    debit

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12043 entries, 0 to 12042
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   status             12043 non-null  object        
 1   card_present_flag  7717 non-null   float64       
 2   bpay_biller_code   885 non-null    object        
 3   account            12043 non-null  object        
 4   currency           12043 non-null  object        
 5   long_lat           12043 non-null  object        
 6   txn_description    12043 non-null  object        
 7   merchant_id        7717 non-null   object        
 8   merchant_code      883 non-null    float64       
 9   first_name         12043 non-null  object        
 10  balance            12043 non-null  float64       
 11  date               12043 non-null  datetime64[ns]
 12  gender             12043 non-null  object        
 13  age                12043 non-null  int64         
 14  mercha

From the above info we can conclude there are 12043 transations data for the three months and 23 attributes for each transaction. 

In [6]:
df.isnull().sum()

status                   0
card_present_flag     4326
bpay_biller_code     11158
account                  0
currency                 0
long_lat                 0
txn_description          0
merchant_id           4326
merchant_code        11160
first_name               0
balance                  0
date                     0
gender                   0
age                      0
merchant_suburb       4326
merchant_state        4326
extraction               0
amount                   0
transaction_id           0
country                  0
customer_id              0
merchant_long_lat     4326
movement                 0
dtype: int64

In [7]:
df.groupby(by="txn_description").count()

status  card_present_flag  bpay_biller_code  account  \
txn_description                                                         
INTER BANK          742                  0                 0      742   
PAY/SALARY          883                  0               883      883   
PAYMENT            2600                  0                 0     2600   
PHONE BANK          101                  0                 0      101   
POS                3783               3783                 1     3783   
SALES-POS          3934               3934                 1     3934   

                 currency  long_lat  merchant_id  merchant_code  first_name  \
txn_description                                                               
INTER BANK            742       742            0              0         742   
PAY/SALARY            883       883            0            883         883   
PAYMENT              2600      2600            0              0        2600   
PHONE BANK            101       101            0              0         101   
POS                  3783      3783         3783              0        3783   
SALES-POS            3934      3934         3934              0        3934   

                 balance  date  gender   age  merchant_suburb  merchant_state  \
txn_description                                                                 
INTER BANK           742   742     742   742                0               0   
PAY/SALARY           883   883     883   883                0               0   
PAYMENT             2600  2600    2600  2600                0               0   
PHONE BANK           101   101     101   101                0               0   
POS                 3783  3783    3783  3783             3783            3783   
SALES-POS           3934  3934    3934  3934             3934            3934   

                 extraction  amount  transaction_id  country  customer_id  \
txn_description                                                             
INTER BANK              742     742             742      742          742   
PAY/SALARY              883     883             883      883          883   
PAYMENT                2600    2600            2600     2600         2600   
PHONE BANK              101     101             101      101          101   
POS                    3783    3783            3783     3783         3783   
SALES-POS              3934    3934            3934     3934         3934   

                 merchant_long_lat  movement  
txn_description                               
INTER BANK                       0       742  
PAY/SALARY                       0       883  
PAYMENT                          0      2600  
PHONE BANK                       0       101  
POS                           3783      3783  
SALES-POS                     3934      3934

- From the above results, we can confirm that there are lot of null value columns in the dataset but few columns like card_present_flag, merchant_id, merchant_suburb, merchant_state, merchant_long_lat can just be replaced with zero as these transanctions are not related to POS, SALES-POS but belong to PAYMENT, INTER BANK, PAY/SALARY or PHONE BANK where merchant is not involved.
- And we can observe that bpay_biller_code, merchant_code is given only to PAY/SALARY related transactions.

In [8]:
df = df.fillna(0)

In [9]:
#confirming there are no null values in the dataset
df.isnull().sum()

status               0
card_present_flag    0
bpay_biller_code     0
account              0
currency             0
long_lat             0
txn_description      0
merchant_id          0
merchant_code        0
first_name           0
balance              0
date                 0
gender               0
age                  0
merchant_suburb      0
merchant_state       0
extraction           0
amount               0
transaction_id       0
country              0
customer_id          0
merchant_long_lat    0
movement             0
dtype: int64

In [10]:
df.to_csv(r"clean_data.csv",index=False)

In [11]:
df["date"].nunique()

91

__Time series of daily transactions__

![trans_count_daily](trans_count_daily.PNG)

The above graph shows that theris clear seasonalityin the transactions.

In [124]:
day_top_ten = df["date"].value_counts().head(3)
day_last_ten = df["date"].value_counts().tail(3)

In [129]:
import datetime
for count, day in zip(day_top_ten,day_top_ten.index.tolist()):
    print(str(day.date()) + " | "+ str(count)+" | "+ day.strftime("%A"), )

2018-09-28 | 174 | Friday
2018-08-17 | 172 | Friday
2018-10-05 | 168 | Friday


In [130]:
for count, day in zip(day_last_ten,day_last_ten.index.tolist()):
    print(str(day.date()) + " | "+ str(count)+" | "+ day.strftime("%A"), )

2018-10-23 | 96 | Tuesday
2018-10-08 | 95 | Monday
2018-10-30 | 89 | Tuesday
